box: implementation of rtree pagination

Vladimir1t opened the pull request:
Implementation of rtree pagination for neighbor iterator.

for example:
```
tuples, pos = rtree_index:select ({1, 1}, {iterator = 'neighbor', fetch_pos=true, limit=6})
tuples, pos = rtree_index:select ({1, 1}, {iterator = 'neighbor', fetch_pos=true, limit=3, after=pos})
```

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,90 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
 	/** Memory pool the iterator was allocated from. */
 	struct mempool *pool;
+	/* last fetched tuple (pos). Is used for pagination */
+	tuple* last;
```
Please use `struct tuple *last`.

Vladimir1t replied:
fixed

drewdzzz commented on src/box/index.cc:
```diff
@@ -749,6 +749,9 @@ iterator_position_validate(const char *pos, uint32_t pos_part_count,
 	if (key_validate_parts(cmp_def, pos, pos_part_count, true, &pos_end) != 0)
 		goto fail;
+	/* Can skip, if it is rtree index. Is used for rtree pagination */
+	if (type == ITER_NEIGHBOR)
```
Let's try to pass `index_def` instead of `cmp_def` here so we could use `index_type`. And call `key_compare` only if the index has TREE type.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -176,7 +246,12 @@ index_rtree_iterator_next(struct iterator *i, struct tuple **ret)
 /********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
 		memtx_tx_story_gc();
 /*********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND END**********/
+
```
Please remove unnecessary line.

Vladimir1t replied:
removed, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -176,7 +246,12 @@ index_rtree_iterator_next(struct iterator *i, struct tuple **ret)
 	} while (*ret == NULL);
+	/* Save last tuple. Is used for pagination */
+	if (itr->impl.op == SOP_NEIGHBOR)
+		itr->last = *ret;
```
Please use `tuple_ref` here so that the tuple won't be deleted. And do not forget to call `tuple_unref` in `index_rtree_iterator_free`.

Vladimir1t replied:
used, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -202,7 +277,6 @@ memtx_rtree_index_def_change_requires_rebuild(struct index *index,
 	    index->def->opts.dimension != new_def->opts.dimension)
 		return true;
 	return false;
-
```
Unnecessary change.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +378,57 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
 }
 
+/**
+ * Implementation of iterator position for general and multikey indexes.
```
Please rephrase - there are no multi key RTREE indexes.

Vladimir1t replied:
fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +378,57 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
+	assert(def != NULL);
+	assert(pos != NULL);
+
+	struct tuple *tuple = last;
```
Why do you need this variable at all? Please, drop it.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +378,57 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
+    * memory allocation will not fail during any operation
+    * on rtree, because there is no error handling in the
+    * rtree lib.
+    */
```
Please fix the indentation.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -388,11 +499,27 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
+							index, &rect);
+		} else {
+			diag_set(UnsupportedIndexFeature, base->def,
+				 "pagination with wrong iterator type");
```
Please, provide more details in error.
Also, let's put this error to the old place so that it would be raised as early as possible.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -388,11 +499,27 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
 	}
+
+	if (op == SOP_NEIGHBOR) {
+		it->base.position = rtree_iterator_position;
```
Please, move the whole logic after `iterator_create` call.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -388,11 +499,27 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
 	it->pool = &memtx->rtree_iterator_pool;
 	it->base.next_internal = index_rtree_iterator_next;
 	it->base.next = memtx_iterator_next;
-	it->base.position = generic_iterator_position;
```
Why did you remove this line? You should still set the method for non-NEIGHBOR iterators.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,90 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
+ * Euclid distance.
+ */
+static sq_coord_t
+rtree_rect_neigh_distance_pos(const struct rtree_rect *rect,
```
We should implement Manhattan distance as well.

Vladimir1t replied:
thanks, fixed

drewdzzz reviewed (changes requested):
Thanks for your patch! I've left some comments, and more to come later.
Also, please, fix the linters.

Vladimir1t commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,90 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
-        struct iterator base;
-        struct rtree_iterator impl;
+    struct iterator base;
+    struct rtree_iterator impl;
```
fixed

Vladimir1t commented on src/box/index.cc:
```diff
@@ -749,6 +749,9 @@ iterator_position_validate(const char *pos, uint32_t pos_part_count,
 	const char *pos_end;
```
thanks, fixed

xuniq commented:
I don't see any changes in the changelog file. Please add a changelog entry for this feature.

drewdzzz commented on src/box/box.cc:
```diff
@@ -3869,7 +3869,7 @@ box_iterator_position_pack(const char *pos, const char *pos_end,
 box_iterator_position_unpack(const char *packed_pos,
 			     const char *packed_pos_end,
-			     struct key_def *cmp_def, const char *key,
+			     struct index_def *ind_def, const char *key,
```
Let's call it `index_def`.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/index.cc:
```diff
@@ -740,22 +740,26 @@ iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
-		goto fail;
+	/* Can skip, if it is rtree index. */
+	if (ind_def->type == TREE) {
+		/* Position msut meet the search criteria. */
```
Please, fix the typo: `msut` -> `must`. Yes, it was here before you, let's just fix it along the way.

drewdzzz commented on src/box/index.cc:
```diff
@@ -740,22 +740,26 @@ iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
-		goto fail;
-	if ((type == ITER_EQ || type == ITER_REQ) && cmp != 0)
-		goto fail;
+	/* Can skip, if it is rtree index. */
```
Please, drop the comment.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/index.h:
```diff
@@ -416,7 +416,7 @@ iterator_position_unpack(const char *packed_pos, const char *packed_pos_end,
 iterator_position_validate(const char *pos, uint32_t pos_part_count,
 			   const char *key, uint32_t key_part_count,
-			   struct key_def *cmp_def, enum iterator_type type);
+			   struct index_def *ind_def, enum iterator_type type);
```
Ditto.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
+	struct rtree_iterator impl;
 	/** Memory pool the iterator was allocated from. */
 	struct mempool *pool;
+	/* last fetched tuple (pos). Is used for pagination */
```
Please, start with a capital letter and finish with a dot : `/* A comment. */`
Also, what does `pos` mean here?
Also, please add the second `*` symbol at the beginning of the comment.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
+	struct tuple *last;
 };
 
+/* Min distance of last saved tuple.
```
Please, fix this comment and all the other ones. How we write them:
One-liners:
```c
/** The comment. */
```

Multi-liners:
```c
/**
 * Multi line
 * comment.
 */
```

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
+ * Euclid distance.
+ */
+static sq_coord_t
+rtree_rect_neigh_distance_pos(
```
Let's call them Euclid and Manhattan.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
+	for (int i = dimension; --i >= 0; ) {
+		const coord_t *coords = &rect->coords[2 * i];
+		coord_t neigh_coord = neigh_rect->coords[2 * i];
+		if (neigh_coord < coords[0]) {
```
Let's factor this logic of choosing diff out of the function. The function would return `diff` and flag `is_in_rectangle`.

Also, please, cover the new function with an explanation of what it does.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
 	mempool_free(itr->pool, itr);
 }
 
+/* set position to the iterator NEIGHBOR. Is used for pagination */
```
Ditto.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -177,6 +296,12 @@ index_rtree_iterator_next(struct iterator *i, struct tuple **ret)
 		memtx_tx_story_gc();
 /*********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND END**********/
 	} while (*ret == NULL);
+	/* Save last tuple. Is used for pagination */
```
Please, end the comment with the dot.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +428,56 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
+		*size = 0;
+		return 0;
+	}
+	int mk_idx = MULTIKEY_NONE;
```
Why don't you pass MULTIKEY_NONE to the `tuple_extract_key` directly?

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +428,56 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
 
+/*
+ * Implementation of iterator position for neighbor iterator.
+ * Is used for pagination
```
Again, dot at the end.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,15 +428,56 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
-         * on rtree, because there is no error handling in the
-         * rtree lib.
-         */
+	 * In case of rtree we use reserve to make sure that
```
Let's drop the change - you don't touch the function at all.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -332,11 +497,11 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
+			 "pagination with wrong iterator type."
+			 "Only for neighbor iterator is supported");
 	}
-
```
Let's return the newline back.

drewdzzz commented on test/engine-luatest/pagination_test.lua:
```diff
@@ -924,23 +923,3 @@ no_sup.after_each(function(cg)
-                box.space.s.index.sk.select, box.space.s.index.sk,
-                nil, {fullscan=true, fetch_pos=true})
-        local tuple = {0, 0}
-        if index_type == 'rtree' then
-            tuple = {0, {0, 0}}
-        end
```
Why did you drop the test completely? Please, drop only these lines about `rtree`.

drewdzzz commented on test/engine-luatest/pagination_test.lua:
```diff
@@ -895,8 +895,7 @@ end
             {engine = 'memtx', type = 'hash'},
-            {engine = 'memtx', type = 'bitset'},
-            {engine = 'memtx', type = 'rtree'},
+            {engine = 'memtx', type = 'bitset'}
```
Please, return the comma back.

drewdzzz commented on test/box-luatest/rtree_pagination_test.lua:
```diff
@@ -0,0 +1,318 @@
+    end)
+end
+
+g.test_rtree_pagination_neighbor_points = function()
```
Please, use different names for all tests. Otherwise, one of them won't run at all.

drewdzzz commented on test/box-luatest/rtree_pagination_test.lua:
```diff
@@ -0,0 +1,318 @@
+
+        t.assert_equals(tuples, {{1741,{4,1,7,1,2,5,-14,2}}})
+    end)
+end
```
Please, add a new test scenario:
1. Fill space with a lot of points equally distanced from a key.
2. Then check if order of pagination is still stable.

Vladimir1t replied:
thanks, fixed

drewdzzz commented on test/engine-luatest/pagination_test.lua:
```diff
@@ -895,8 +895,7 @@ end
 local no_sup = t.group('Unsupported pagination', {
             {engine = 'memtx', type = 'hash'},
-            {engine = 'memtx', type = 'bitset'},
-            {engine = 'memtx', type = 'rtree'},
```
We still need a test to check if all iterator types except for NEIGHBOR do not support pagination.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -332,9 +507,10 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
+	if (pos != NULL && type != ITER_NEIGHBOR) {
+		diag_set(UnsupportedIndexFeature, base->def,
+			 "pagination with wrong iterator type. "
+			 "Only for neighbor iterator is supported");
```
You don't return `NULL` here, hence, the error is not thrown.
Please, fix, and cover with a test.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -304,6 +438,47 @@ memtx_rtree_index_replace(struct index *base, struct tuple *old_tuple,
+		index_weak_ref_get_index_checked(&it->index_ref);
+	struct index_rtree_iterator *rtree_it =
+		(struct index_rtree_iterator *)it;
+	return rtree_iterator_position_impl(
```
There is no need in separate `impl` function. Just inline it right here.

drewdzzz commented on src/lib/salad/rtree.h:
```diff
@@ -116,6 +120,13 @@ enum rtree_distance_type {
+enum position_type {
+	NO_PAGINATION = 0, /* No rtree pagination */
+	SET_POS = 1, /* new position set */
+	FIND_POS = 2 /* find pos tuple in rtree */
```
Why do you need it?
AFAICS, `FIND_POS` is not needed at all - you only use it once, and that usage has no effect.

drewdzzz replied:
Please, address this comment.

drewdzzz commented on src/lib/salad/rtree.c:
```diff
@@ -916,9 +984,26 @@ rtree_iterator_next(struct rtree_iterator *itr)
 				return NULL;
 			rtnt_remove(&itr->neigh_tree, neighbor);
+
+			/* Pagination optimisaion */
```
Pagination is about starting an iterator in a particular place. So let's implement it completely in `rtree_search`. Also, we wouldn't need to save anything to the iterator with this approach - just pass the pagination options right to `rtree_search`.

drewdzzz commented on src/box/memtx_rtree.cc:
```diff
@@ -403,6 +577,7 @@ memtx_rtree_index_create_iterator(struct index *base, enum iterator_type type,
 	 * caller's side.
 	 */
 	rtree_search(&index->tree, &rect, op, &it->impl);
+	// rtree_skip_neighbors(pos, &rect, &index->tree, &it->impl);
```
Let's call this function here, not in `rtree_search`.

drewdzzz commented on test/box-luatest/rtree_pagination_test.lua:
```diff
@@ -0,0 +1,479 @@
+local server = require('luatest.server')
```
Please add tests where:
1. Tuple which position is used was deleted from space.
2. New tuples are inserted before reading each next page.

drewdzzz commented on src/lib/salad/rtree.h:
```diff
@@ -32,6 +32,7 @@
  */
 #include <stddef.h>
 #include <stdbool.h>
+#include "box/index.h"
```
That's not right. `box` is another layer, so you shouldn't use it at all, including `index`. It should be done with a callback.

drewdzzz commented on src/lib/salad/rtree.h:
```diff
@@ -53,24 +54,33 @@ typedef double area_t;
 #if defined(__cplusplus)
 extern "C" {
 #endif /* defined(__cplusplus) */
+/* A box in RTREE_DIMENSION space */
```
Why did you move it here?!

drewdzzz commented on src/lib/salad/rtree.h:
```diff
@@ -53,24 +54,33 @@ typedef double area_t;
+};
+
+/* Coords */
+struct rtree_rect {
```
Why did you move it here?!

drewdzzz commented on src/lib/salad/rtree.h:
```diff
@@ -53,24 +54,33 @@ typedef double area_t;
 	int level;
 	sq_coord_t distance;
+	/* Index. Is used for sorting neighbors with same distence */
+	struct index *index;
```
As I already said, we are not allowed to use `index` here since it's another layer. Let's use a callback.

drewdzzz commented on src/lib/salad/rtree.c:
```diff
@@ -1156,7 +1299,7 @@ rtree_debug_print_page(const struct rtree *tree, const struct rtree_page *page,
 	printf("%d:\n", path);
 	unsigned d = tree->dimension;
-	for (int i = 0; i < page->n; i++) {
+	for (unsigned i = 0; i < page->n; i++) {
```
Why did you change it? You don't need this change in your patch.

drewdzzz commented on src/lib/salad/rtree.c:
```diff
@@ -1096,18 +1238,19 @@ rtree_search(const struct rtree *tree, const struct rtree_rect *rect,
 			struct rtree_rect cover;
 			rtree_page_cover(tree, tree->root, &cover);
 			sq_coord_t distance;
-			if (tree->distance_type == RTREE_EUCLID)
+			if (tree->distance_type == RTREE_EUCLID) {
 				distance =
-				rtree_rect_neigh_distance2(&cover, rect,
-							   tree->dimension);
-			else
+				rtree_rect_neigh_distance2(
+					&cover, rect, tree->dimension);
+			} else { /* RTREE_MANHATTAN */
 				distance =
-				rtree_rect_neigh_distance(&cover, rect,
-							  tree->dimension);
+				rtree_rect_neigh_distance(
+					&cover, rect, tree->dimension);
+			}
 			struct rtree_neighbor *n =
-				rtree_iterator_new_neighbor(itr, tree->root,
-							    distance,
-							    tree->height);
+				rtree_iterator_new_neighbor(
+					itr, tree->root,
+					distance, tree->height);
```
You don't need these changes, they are purely cosmetic. Please, drop them.

drewdzzz commented on src/lib/salad/rtree.c:
```diff
@@ -874,20 +972,21 @@ rtree_iterator_process_neigh(struct rtree_iterator *itr,
 		struct rtree_page_branch *b;
 		b = rtree_branch_get(itr->tree, pg, i);
 		coord_t distance;
-		if (itr->tree->distance_type == RTREE_EUCLID)
-			distance = rtree_rect_neigh_distance2(&b->rect,
-							      &itr->rect, d);
-		else
-			distance = rtree_rect_neigh_distance(&b->rect,
-							     &itr->rect, d);
+		if (itr->tree->distance_type == RTREE_EUCLID) {
+			distance = rtree_rect_neigh_distance2(
+				&b->rect, &itr->rect, d);
+		} else { /* RTREE_MANHATTAN */
+			distance = rtree_rect_neigh_distance(
+				&b->rect, &itr->rect, d);
+		}
 		struct rtree_neighbor *neigh =
-			rtree_iterator_new_neighbor(itr, b->data.page,
-						    distance, level - 1);
+			rtree_iterator_new_neighbor(
+				itr, b->data.page, distance,
+				level - 1);
```
Again, you don't need these changes - they are purely cosmetic. Let's remove them.

drewdzzz reviewed (changes requested):
Thank you for your work!
I've left a bunch of comments to make the patch prettier.

AArdeev commented:
Please add changelog item @Vladimir1t

Vladimir1t committed:
box: sorting neighbors with same distance

If distances of two rects are equal, they will be sorted according to primary key.
Add new callback 'tie_cmp_cb' for using it in 'neighbor_cmp'

NO_DOC=bugfix
NO_CHANGELOG=later

Vladimir1t committed:
box: fix problem in index.cc in iterator_position_validate

Made this checking only for Tree index.
Change key_def into index_def

NO_DOC=bugfix
NO_CHANGELOG=later

Vladimir1t committed:
box: implementation of rtree pagination

This patch supports pagination of neighbor iterator.
There are 3 algorithms of saving, decoding position and
iterating after current posiotion.
Implement options 'after' and 'fetch_pos' in select method of iterator.
If 'after' is specifiend it must contain a tuple from which selection
must continue or its position.

For rtree index (memtx) select internally skip all tuples before then
pos_tuple in order of neighbor iterator of rtree index.
Pagination with after option is more effective that limit-offset
(distence < pos_distence) approach and it doesn't depend
on the number of previously selected tuples.
This optimisation is implemented in 'rtree.c' in 'rtree_skip_neighbors_equal_distance'.

Add new spaces in 'struct index_rtree_iterator',
'struct rtree_neighbor' for saving position.

In IPROTO tuple is MP_ARRAY, position is MP_BIN.

Example of using:
```
tuples, pos = rtree_index:select({1, 1},
		{iterator = 'neighbor',
           	fetch_pos=true, limit=4})

tuples = rtree_index:select({1, 1},
		{iterator = 'neighbor',
            	fetch_pos=false, limit=2,
		after=pos})
```

The same issue for tree index: #7639
NO_DOC=bugfix
NO_CHANGELOG=later

lenkis reviewed (commented):
Please add a changelog entry for this PR.
